src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java

```diff
@@ -47,29 +48,104 @@ public void setCfg(final VIPCfg cfg) {
     public TranslationMessage() {
         super();
     }
-
+    
+    /**
+     * Retrieves the localized message from the cache
+     * 
+     * @param component The Singleton component in which the message belongs
+     * @param key The key that represents the message
+     * @param locale The locale in which the message is requested to be localized
+     * @return One of the items in the following priority-ordered list: 
+     * <ul>
+     * 		<li>The message in the requested locale</li> 
+     * 		<li>The message in the default locale</li>
+     * 		<li>null</li>
+     * </ul>
+     */
+    private String getCachedMessage(String component, String key, Locale locale) {
+    	MessagesDTO dto = new MessagesDTO(component, key, null, locale.toLanguageTag(), this.cfg);
+    	StringService s = new StringService();
+    	return s.getString(dto);
+    }
+    
+    
+    /**
+     * Retrieves the localized message from the cache, with added functionality such as:
+     * <ul>
+     * 	<li>Pseudo-localization</li>
+     * 	<li>Fallback to source message when message is neither collected nor translated yet</li>
+     * </ul>
+     * 
+     * @param locale The locale in which the message is requested to be localized
+     * @param component The Singleton component in which the message belongs
+     * @param sourceOpt The optional SourceOpt object which gives access to source messages
+     * @param key The key that represents the message
+     * @param args Values to replace placeholders in the message with
+     * @return One of the items in the following priority-ordered list: 
+     * <ul>
+     * 		<li>The pseudo message, if isPseudo is true</li> 
+     * 		<li>The message in the requested locale, if available</li>
+     * 		<li>The message in the default locale, if available</li>
+     * 		<li>The message from sourceOpt, if available</li>
+     * 		<li>key</li>
+     * </ul>
+     */
+    public String getMessage(final Locale locale, final String component, final String key, final Object... args) {
+    	String message = null;
+    	SourceOpt sourceOpt = VIPCfg.getInstance().getSrcOpt();
+    	String source = (sourceOpt == null) ? null : sourceOpt.getMessage(key);
+    	
+    	// If sourceOpt is defined, pseudo-translation which uses the source message is supported
+    	if (VIPCfg.getInstance().isPseudo() && source != null && sourceOpt.getLocale().equals(locale)) {
+    		return ConstantsKeys.PSEUDOCHAR2 + FormatUtils.format(source, sourceOpt.getLocale(), args) + ConstantsKeys.PSEUDOCHAR2;
+    	}
+    	
+    	// Get the message in the target locale
+    	message = getCachedMessage(component, key, locale);
```

Comment:
if the message is not empty, should return it and skip source handling?

Reply:
That was my initial thought, but the code in getString uses the source message for the following:
    ```
	// 	a. if neither localized message nor default locale message was not retrieved successfully
    	// 	b. for pseudo-translation
    	//      c. if the message hasn't been collected for localization
```
Can we skip these?